From 49048aab9a4637c923abbfb8b74e73dc14da41c4 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Sun, 05 Jun 2022 15:47:47 +0300
Subject: [PATCH] Revert "[SW-86865] gaudi3: update fit run image routine"

Revert rom recovery implementation for secured firmware update via
inband and OOB.

This reverts commit 9bb2b53c4c7dd8260288efe475436fcfc825fee4.

Change-Id: I087fee8dabe854c7575eb158b32a969582410e9f
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 0d51f53..9682b1d 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -898,22 +898,17 @@
 	regmap_writel(0x0, run_halt_offset);
 }
 
-void arcboot_start_image(uint64_t loadaddr, int core_id)
+void arcboot_start_image(uint64_t startaddr)
 {
 	void (*image_entry)(void);
 	uint32_t arc_addr;
 
-	if (!loadaddr)
+	if (!startaddr)
 		wfe();
 
-	if (core_id != FIT_CORE_ID_D0_PSOC_ARC1) {
-		run_halt_arc(core_id, 1);
-		return;
-	}
+	arc_addr = (uint32_t)arcboot_rom_get_arc_addr(startaddr);
 
-	arc_addr = (uint32_t)arcboot_rom_get_arc_addr(loadaddr);
-
-	image_entry = (void *) (*(uint32_t *)arc_addr);
+	image_entry = (void *)((uint32_t) arc_addr);
 	image_entry();
 }
 
diff --git a/common/fit.c b/common/fit.c
index 10aef8c..ca463fd 100644
--- a/common/fit.c
+++ b/common/fit.c
@@ -802,30 +802,24 @@
 
 void fit_run_image(void)
 {
-	uint64_t loadaddr = 0;
+	uint64_t startaddr = 0;
 	uint32_t i;
-	bool run_psoc_arc1 = false;
 
 	for (i = 0; i < FIT_MAX_IMAGES; i++) {
 
-		if (!images[i].arch)
-			break;
-
-		/* start image of ARC1 image at last */
-		if ((images[i].core_id == FIT_CORE_ID_D0_PSOC_ARC1) && (!run_psoc_arc1)) {
-			run_psoc_arc1 = true;
-			loadaddr = images[i].loadaddr;
+		if (strcmp(images[i].arch, "arc"))
 			continue;
-		}
-
-		if (!arc_info[images[i].core_id].run) {
-			arcboot_start_image(images[i].loadaddr, images[i].core_id);
-			arc_info[images[i].core_id].run = true;
-		}
+		/*
+		 * Since we only perform agent fit download as part
+		 * of ROM recovery, assumption is made here that
+		 * number of fit images will be 1 and agent will
+		 * run only on ARC.
+		 */
+		startaddr = images[i].startaddr;
+		break;
 	}
 
-	if (run_psoc_arc1)
-		arcboot_start_image(loadaddr, FIT_CORE_ID_D0_PSOC_ARC1);
+	arcboot_start_image(startaddr);
 }
 
 __weak size_t arcboot_get_img_sec_desc_size(void)
diff --git a/include/arcboot.h b/include/arcboot.h
index 264db56..8044169 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -290,7 +290,7 @@
 int arcboot_rom_mem_init(uint64_t addr, uint32_t size);
 void arcboot_rom_verify(void);
 bool arcboot_is_valid_address_range(uint64_t addr, int len);
-void arcboot_start_image(uint64_t loadaddr, int core_id);
+void arcboot_start_image(uint64_t startaddr);
 uint32_t arcboot_error_read_and_clr(uint32_t *major, uint32_t *minor);
 bool arcboot_error_inform(uint32_t major, uint32_t minor);
 bool arcboot_error_is_pending(void);
